src/Avalonia.Controls/Application.cs

```diff
@@ -190,108 +194,163 @@ public IResourceDictionary Resources
+            if (_mainLoopCancellationTokenSource != null)
             {
-                if (mainWindow == null)
-                {
-                    throw new ArgumentNullException(nameof(mainWindow));
-                }
+                throw new InvalidOperationException("Application is already running.");
+            }
 
-                if (!mainWindow.IsVisible)
-                {
-                    mainWindow.Show();
-                }
+            _mainLoopCancellationTokenSource = tokenSource;
 
-                MainWindow = mainWindow;
-            }           
+            Dispatcher.UIThread.Post(() => OnStartup(new StartupEventArgs()), DispatcherPriority.Send);
 
             Dispatcher.UIThread.MainLoop(_mainLoopCancellationTokenSource.Token);
 
-            // Make sure we call OnExit in case an error happened and Exit() wasn't called explicitly
-            if (!IsExiting)
+            if (!IsShuttingDown)
             {
-                OnExit?.Invoke(this, EventArgs.Empty);
+                Shutdown(_exitCode);
             }
+
+            return _exitCode;
         }
 
         /// <summary>
-        /// Runs the application's main loop until the <see cref="CancellationToken"/> is canceled.
+        /// Raises the <see cref="Startup"/> event.
         /// </summary>
-        /// <param name="token">The token to track</param>
-        public void Run(CancellationToken token)
+        /// <param name="e">A <see cref="StartupEventArgs"/> that contains the event data.</param>
+        protected virtual void OnStartup(StartupEventArgs e)
         {
-            Dispatcher.UIThread.MainLoop(token);
-
-            // Make sure we call OnExit in case an error happened and Exit() wasn't called explicitly
-            if (!IsExiting)
-            {
-                OnExit?.Invoke(this, EventArgs.Empty);
-            }
+            Startup?.Invoke(this, e);
         }
 
         /// <summary>
-        /// Exits the application
+        /// Raises the <see cref="Exit"/> event.
         /// </summary>
-        public void Exit()
+        /// <param name="e">A <see cref="ExitEventArgs"/> that contains the event data.</param>
+        protected virtual void OnExit(ExitEventArgs e)
         {
-            IsExiting = true;
+            Exit?.Invoke(this, e);
+        }
+
+        /// <inheritdoc/>
+        public void Shutdown(int exitCode = 0)
+        {
+            if (IsShuttingDown)
+            {
+                throw new InvalidOperationException("Application is already shutting down.");
+            }
+
+            _exitCode = exitCode;
+
+            IsShuttingDown = true;         
 
             Windows.Clear();
 
-            OnExit?.Invoke(this, EventArgs.Empty);
+            try
+            {
+                var e = new ExitEventArgs { ApplicationExitCode = _exitCode };
+
+                OnExit(e);
+
+                _exitCode = e.ApplicationExitCode;
+
+                Environment.ExitCode = _exitCode;
```

Comment:
Also `_exitCode`, but `IsShuttingDown`

Reply:
Exit is the point in time when everything stops. Shutting down is a range in time. They don't need to be named the same because they are not the same.

Reply:
This is consistent with WPF:

https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.shutdown?view=netframework-4.8
https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.onexit?view=netframework-4.8
https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.exit?view=netframework-4.8